glommio/src/executor/multitask.rs

```diff
@@ -110,11 +114,32 @@ impl LocalQueue {
     }
 }
 
+const MAX_UNFAIR_SCHEDULING: usize = 256;
+
 /// A single-threaded executor.
 #[derive(Debug)]
 pub(crate) struct LocalExecutor {
     local_queue: LocalQueue,
 
+    /// run_next, if present is a runnable that was scheduled by the previous
```

Comment:
aren't we injecting this latency into the task that was sitting in the queue, though?

If you have a task that was sitting in the queue ready to run (let's say the result of some I/O that finally came back), and now you have a chain of 256 tasks that will hit the mini-LIFO, all we've done was move latency around, no?

This is not to say the change is bad. A newly created task running immediately tends to be good because it will dispatch I/O earlier.

I'd like to think a bit more about adversarial cases and what kind of workloads triggers them.

You said this technique is used by the Go scheduler, so maybe they have some write up on this ?

Reply:
I'm reasonably comfortable with the go's runtime code so that was my guide, you can find it here: https://cs.opensource.google/go/go/+/master:src/runtime/runtime2.go;l=627-639;drc=master;bpv=1;bpt=1  It is very well documented.

Additionally, this is the commit that introduced the mechanism in Go: https://cs.opensource.google/go/go/+/e870f06c3f49ed63960a2575e330c2c75fc54a34

Reply:
The idea is to run a task's continuations as deep as possible to take advantage of data locality. As you said, the usual case is that we stop doing so well before reaching the limit of 256 because the task will return `Poll::Pending` because of IO or some other data dependencies.

As for starving other tasks in the queue, I share your concern and the limit of 256 was added to limit this. Go doesn't need it because it implements per-goroutine runtime quotas (a goroutine is a `Task` in Glommio). We have a limit per task queue instead.

> If you have a task that was sitting in the queue ready to run (let's say the result of some I/O that finally came back), and now you have a chain of 256 tasks that will hit the mini-LIFO, all we've done was move latency around, no?

Yes, up to the latency limit that's attached to the task queue. This system doesn't prevent the scheduler from yielding if necessary (`need_preempt()` is still called in between each time). 

So, fundamentally, the before/after situation hasn't changed. If you define a task queue with a latency requirement of 1ms and you spawn 1000 tasks inside, the scheduler guarantees that the queue (and not the tasks inside) will be visited a minimum of once per ~millisecond. So if each of those tasks are CPU hogs, then they will be visited about once per second at most.